### PR TITLE
message streams to recieve events from outside of the bot and send to any channel

### PR DESCRIPTION
**Pull request by bnfinet**
see discussion in #97 

 * allows registration of MessageStream functions
 * routes any received Message to the specified ChannelData Channel/Server/Protocol
 * adds config items Protocol and Server to each bot instance
 * adds test for MessageStreams to cmd_tests.go

### CODE REVIEW

File: cmd.go [hunk 1]
```diff
@@ -358,3 +417,38 @@ func (b *Bot) checkCmdError(err error, c *Cmd) {
...
+// when a message is sent into a chan it has a good chance of arriving at the wrong bot instance
+// for every message we check to see if it matched this b.Protocol and b.Server
+// if it doesn't we lookup the entry in messageStreams[key] send it to *that* Data chan
+func (b *Bot) handleMessageStream(streamName string, ms *MessageStream) {
+	for {
+		select {
+		case d := <-ms.Data:
+
+			if d.ChannelData.Protocol != b.Protocol || d.ChannelData.Server != b.Server {
+				// b.errored("Warning: MessageStream "+cmdName+" for "+d.ChannelData.Channel+" will not be received by "+b.Server, ErrProtocolServerMismatch)
```
Comment (fabioxgn) [inline cmd.go]: Forgot to remove this line?

File: cmd_test.go [hunk 2]
```diff
@@ -672,3 +686,68 @@ func TestFilterCommandSilence(t *testing.T) {
 		t.Error("Expected 1 error")
 	}
 }
+
+// how to test channels..
+// https://www.hugopicado.com/2016/10/01/testing-over-golang-channels.html
```
Comment (fabioxgn) [inline cmd_test.go]: 👏

File: telegram/telegram.go [hunk 3]
```diff
@@ -47,13 +52,20 @@ func Run(token string, debug bool) {
...
 	b := bot.New(&bot.Handlers{
 		Response: responseHandler,
-	})
+	}, &bot.Config{
+		Protocol: protocol,
+		Server:   server,
+	},
+	)
+	// b.Protocol = protocol
+	// b.Server = server
```
Comment (fabioxgn) [inline telegram/telegram.go]: Can you remove this, please?

File: cmd.go [hunk 4]
```diff
@@ -73,6 +79,27 @@ type User struct {
...
+	Message     string
+	ChannelData *ChannelData
+}
+
+// MessageStreamConfig set the Protocol, Server and Channel where the messages should be sent
+type MessageStreamConfig struct {
+	Version    int
+	StreamName string
+	MsgFunc    messageStreamFunc
+}
```
[the remark this answers is not in the document]
Comment (bnfinet) [reply cmd.go]: I've unexported the properties and struct

File: bot.go [hunk 5]
```diff
@@ -61,16 +75,46 @@ func New(h *Handlers) *Bot {
...
+				b.errored("MessageStream "+config.StreamName+" failed ", err)
+			}
+			msKey := messageStreamKey{
+				Protocol:   b.Protocol,
+				Server:     b.Server,
+				StreamName: config.StreamName,
+			}
+			// thread safe write
+			msMap.messageStreams[msKey] = ms
+			msMap.Unlock()
```
[the remark this answers is not in the document]
Comment (bnfinet) [reply bot.go]: somehow, when using `defer nsMap.Unlock()` the tests fail.  I'm keeping this as is for the time being but I'd be curious as to why `defer` is not working here

**Pull request merged**